File: typescript/lib/base/models/rectangle.ts
import { Dimensional } from "../commons";

export class Rectangle implements Dimensional {
    private width: number;
    private height: number;

    constructor(width: number, height: number) {
        const dimOne: number = width > 0 ? width : 1;
        const dimTwo: number = height > 0 ? height : 1;
        this.width = dimOne;
        this.height = dimTwo;
    }

    area(): number {
        return this.width * this.height;
    }

    perimeter(): number {
        return 2 * (this.width + this.height);
    }
}

File: rust/src/base/models/rectangle.rs
use crate::base::commons::Dimensional;

pub struct Rectangle {
    pub width: f64,
    pub height: f64,
}

impl Rectangle {
    pub fn new(width: f64, height: f64) -> Self {
        let dim_one: f64 = if width < 0.0 { 1.0 } else { width };
        let dim_two: f64 = if height < 0.0 { 1.0 } else { height };
        Self {
            width: dim_one,
            height: dim_two,
        }
    }
}

impl Dimensional for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn perimeter(&self) -> f64 {
        2.0 * (self.width + self.height)
    }
}
